core/src/main/scala/kanaloa/reactive/dispatcher/metrics/MetricsCollector.scala

```diff
@@ -1,24 +1,199 @@
+            tryComplete(s)
+            continue(QueueStatus(poolSize = Some(size)))
+          }
+      }
+
+    case AddSample ⇒
+      continue(tryComplete(s))
+
+  }
 
-object NoOpMetricsCollector extends MetricsCollector {
-  def send(metric: Metric): Unit = () // Do nothing
+
+  def partialUtilized(poolSize: Option[Int]): Receive = handleSubscriptions orElse {
+    case PoolIdle(size) if size <= thresholdOfIdleWorkers ⇒
+      context become fullyUtilized(QueueStatus(poolSize = poolSize))
+    case PoolIdle(idle) ⇒
+      publishUtilization(idle, poolSize)
+    case metric: Metric ⇒
+      handle(metric) {
+        case PoolSize(s) ⇒
+          context become partialUtilized(Some(s))
+      }
+    case AddSample ⇒
+  }
+
+  private def continue(qs: QueueStatus): Unit = context become fullyUtilized(qs)
+
+  /**
+   *
+   * @param status
+   * @return a reset status if completes, the original status if not.
+   */
+  private def tryComplete(status: QueueStatus): QueueStatus = {
+    status.toSample(minSampleDuration).fold(status) { sample ⇒
+      publish(sample)
+      status.copy(workDone = 0, start = Time.now)
+    }
+  }
+
+  private def handle(metric: Metric)(pf: PartialFunction[Metric, Unit]): Unit = {
+    reporter.foreach(_.report(metric))
+    pf.applyOrElse(metric, (_: Metric) ⇒ ())
+  }
+
+  def publish(report: Report): Unit = {
+    subscribers.foreach(_ ! report)
+  }
 }
 
 object MetricsCollector {
-  /** If statsd config exists, create StatsD, otherwise NoOp */
-  def fromConfig(dispatcherName: String, config: Config)(implicit system: ActorSystem): MetricsCollector = {
-    import net.ceedubs.ficus.Ficus._
-    import net.ceedubs.ficus.readers.ArbitraryTypeReader._
 
-    val defaultSettings = config.as[Option[StatsDMetricsCollectorSettings]]("metrics.statsd").filter(_ ⇒ config.getOrElse("metrics.enabled", true))
+  def props(
+    reporter: Option[Reporter],
+    settings: MetricsCollectorSettings = MetricsCollectorSettings()
+  ): Props = Props(new MetricsCollector(reporter, settings))
+
+  def apply(
+    reporter: Option[Reporter],
+    settings: MetricsCollectorSettings = MetricsCollectorSettings()
+  )(implicit system: ActorSystem): ActorRef = system.actorOf(props(reporter, settings))
 
-    defaultSettings.fold(NoOpMetricsCollector: MetricsCollector)(StatsDMetricsCollector(dispatcherName, _))
+  case object AddSample
+
+  case class Subscribe(actorRef: ActorRef)
+  case class Unsubscribe(actorRef: ActorRef)
+
+  /**
+   *
+   * @param sampleRate
+   * @param minSampleDurationRatio minimum sample duration ratio to sample rate. Sample duration less than this will be abandoned.
+   * @param thresholdOfIdleWorkers The maximum number of idle workers that is allowed for the system to be considered as fully utilized
+   */
+  case class MetricsCollectorSettings(
+    sampleRate:             FiniteDuration = 1.second,
+    minSampleDurationRatio: Double         = 0.3,
+    thresholdOfIdleWorkers: Int            = 0
```

Comment:
Do you think we would be better served by modeling this as a percentage, rather than a fixed number?   2 idleWorkers out of 10 is one thing, but 2 out of ...100?